src/cpu/flexcpu/inflight_inst.cc

```diff
@@ -0,0 +1,895 @@
+    assert(reg_id.isCCReg());
+
+    const DataSource& source = sources[op_idx];
+    const shared_ptr<InflightInst> producer = source.producer.lock();
+
+    if (producer) {
+        // If the producing instruction is still in the buffer, grab the result
+        // assuming that it has been produced, and our index is in bounds.
+        assert(producer->isComplete());
+        assert(source.resultIdx >= 0
+            && source.resultIdx < producer->results.size()
+            && producer->resultValid[source.resultIdx]);
+        return producer->getResult(source.resultIdx).asCCReg();
+    } else {
+        // Either the producing instruction has already been committed, or no
+        // dependency was in-flight at the time that this instruction was
+        // issued.
+        return backingContext->readCCReg(reg_id.index());
+    }
+}
+
+void
+InflightInst::setCCRegOperand(const StaticInst* si, int dst_idx, RegVal val)
+{
+    assert(si->destRegIdx(dst_idx).isCCReg());
+
+    results[dst_idx].set(val, CCRegClass);
+    resultValid[dst_idx] = true;
+}
+
+RegVal
+InflightInst::readMiscRegOperand(const StaticInst* si, int op_idx)
+{
+    const RegId& reg_id = si->srcRegIdx(op_idx);
+    assert(reg_id.isMiscReg());
+
+    const DataSource& source = sources[op_idx];
+    const shared_ptr<InflightInst> producer = source.producer.lock();
+
+    if (producer) {
+        // If the producing instruction is still in the buffer, grab the result
+        // assuming that it has been produced, and our index is in bounds.
+        assert(producer->isComplete());
+        assert(source.resultIdx >= 0
+            && source.resultIdx < producer->results.size()
+            && producer->resultValid[source.resultIdx]);
+        return producer->getResult(source.resultIdx).asMiscReg();
+    } else {
+        // Either the producing instruction has already been committed, or no
+        // dependency was in-flight at the time that this instruction was
+        // issued.
+        return backingContext->readMiscReg(reg_id.index());
+    }
+}
+
+void
+InflightInst::setMiscRegOperand(const StaticInst* si, int dst_idx, RegVal val)
+{
+    assert(si->destRegIdx(dst_idx).isMiscReg());
+
+    results[dst_idx].set(val, MiscRegClass);
+    resultValid[dst_idx] = true;
+}
+
+RegVal
+InflightInst::readMiscReg(int misc_reg)
+{
+    return backingISA->readMiscReg(misc_reg, backingContext);
+}
+
+void
+InflightInst::setMiscReg(int misc_reg, RegVal val)
+{
+    // O3 hides multiple writes to the same misc reg during execution, but due
+    // to potential side effects of each access, I don't think we should be
+    // hiding them. Instead, we will replay each write sequentially during
+    // commit time.
+
+    miscResultIdxs.push_back(misc_reg);
+    miscResultVals.push_back(val);
+    // Values should be applied to the ISA/TC at commit time, when we know the
+    // instruction will not be squashed.
+}
+
+PCState
+InflightInst::pcState() const
```

Comment:
Move to header?

Reply:
I think I left all the stuff that was part of the ExecContext in the source file to keep the definitions together, since they couldn't be in-lined even if put in the header, since they were polymorphic calls, as far as I could tell.